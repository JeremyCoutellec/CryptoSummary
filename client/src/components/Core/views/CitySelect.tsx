import React, { Fragment, useEffect, useState } from 'react';
import { City, State, Country } from 'country-state-city';
import { map, isEqual, filter, includes, debounce, lowerCase, take, join } from 'lodash';

// Material UI
import TextField from '@mui/material/TextField';
import MenuItem from '@mui/material/MenuItem';

const CitySelect = ({ title = '', required = false, onChange = null, defaultValue = '', ...rest }) => {
    const [city, setCity] = useState(defaultValue);
    const [choices, setChoices] = useState([]);
    const allCities = City.getCitiesOfCountry('FR');

    const findCities = debounce(() => {
        let filterCities = [];
        if (city) {
            filterCities = filter(allCities, (cityToFind) => includes(lowerCase(cityToFind.name), lowerCase(city)));
            filterCities = map(filterCities, (cityToFind) => ({
                ...cityToFind,
                country: {
                    ...Country.getCountryByCode(cityToFind?.countryCode),
                },
                state: {
                    ...State.getStateByCodeAndCountry(cityToFind?.stateCode, cityToFind?.countryCode),
                },
            }));
        }
        if (!isEqual(choices, filterCities)) setChoices(filterCities);
    }, 300);

    useEffect(() => {
        findCities();
    }, [city, findCities]);

    const getLabelOfCity = (city) => join([city?.name, city?.state.name, city?.country?.name], ', ');

    return (
        <Fragment>
            <TextField
                variant="outlined"
                margin="normal"
                label={title}
                fullWidth
                required={required}
                value={city}
                onChange={(event) => setCity(event.target.value)}
                {...rest}
            />
            {map(take(choices, 10), (option, index) => (
                <MenuItem
                    key={index}
                    value={option.countryCode}
                    onClick={() => {
                        setCity(getLabelOfCity(option));
                        onChange(option);
                    }}
                >
                    {getLabelOfCity(option)}
                </MenuItem>
            ))}
        </Fragment>
    );
};

export default CitySelect;
