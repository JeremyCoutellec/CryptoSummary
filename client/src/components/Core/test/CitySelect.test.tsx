import React from 'react';
import renderer from 'react-test-renderer';
import CitySelect from '../views/CitySelect';

test('CitySelect rendering', () => {
    const propsCity = {
        title: 'City',
        min: 2,
        max: 50,
        defaultValue: 'Toulouges',
    };
    const component = renderer.create(<CitySelect value={'Toulouges'} {...propsCity} />);
    const citySelect = component.toJSON();
    expect(citySelect).toMatchSnapshot();
});
