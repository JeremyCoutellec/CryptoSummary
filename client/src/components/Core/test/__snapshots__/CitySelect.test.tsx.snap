// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CitySelect rendering 1`] = `
<div
  className="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
  max={50}
  min={2}
>
  <label
    className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink={true}
  >
    City
  </label>
  <div
    className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
      disabled={false}
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      type="text"
      value="Toulouges"
    />
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-1z7n62"
      >
        <span>
          City
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
